File: android/src/main/kotlin/com/solusibejo/flutter_meta_sdk/FlutterMetaSdkMethods.kt
package com.solusibejo.flutter_meta_sdk

import android.os.Bundle
import com.facebook.FacebookSdk
import com.facebook.appevents.AppEventsConstants
import com.facebook.appevents.AppEventsLogger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.util.*

object FlutterMetaSdkMethods {
    fun activateApp(appEventsLogger: AppEventsLogger, result: MethodChannel.Result) {
        appEventsLogger.logEvent(AppEventsConstants.EVENT_NAME_ACTIVATED_APP)
        result.success(null)
    }

    fun clearUserData(result: MethodChannel.Result) {
        AppEventsLogger.clearUserData()
        result.success(null)
    }

    fun setUserData(call: MethodCall, result: MethodChannel.Result) {
        val parameters = call.argument("parameters") as? Map<String, Any>
        val parameterBundle = createBundleFromMap(parameters)

        AppEventsLogger.setUserData(
            parameterBundle?.getString("email"),
            parameterBundle?.getString("firstName"),
            parameterBundle?.getString("lastName"),
            parameterBundle?.getString("phone"),
            parameterBundle?.getString("dateOfBirth"),
            parameterBundle?.getString("gender"),
            parameterBundle?.getString("city"),
            parameterBundle?.getString("state"),
            parameterBundle?.getString("zip"),
            parameterBundle?.getString("country")
        )

        result.success(null)
    }

    fun clearUserId(result: MethodChannel.Result) {
        AppEventsLogger.clearUserID()
        result.success(null)
    }

    fun flush(appEventsLogger: AppEventsLogger, result: MethodChannel.Result) {
        appEventsLogger.flush()
        result.success(null)
    }

    fun getApplicationId(appEventsLogger: AppEventsLogger, result: MethodChannel.Result) {
        result.success(appEventsLogger.applicationId)
    }

    fun getAnonymousId(anonymousId: String, result: MethodChannel.Result) {
        result.success(anonymousId)
    }

    //not an android implementation as of yet
    fun setAdvertiserTracking(result: MethodChannel.Result) {
        result.success(null)
    }

    fun logEvent(appEventsLogger: AppEventsLogger, call: MethodCall, result: MethodChannel.Result) {
        val eventName = call.argument("name") as? String
        val parameters = call.argument("parameters") as? Map<String, Any>
        val valueToSum = call.argument("_valueToSum") as? Double

        if (valueToSum != null && parameters != null) {
            val parameterBundle = createBundleFromMap(parameters)
            appEventsLogger.logEvent(eventName, valueToSum, parameterBundle)
        } else if (valueToSum != null) {
            appEventsLogger.logEvent(eventName, valueToSum)
        } else if (parameters != null) {
            val parameterBundle = createBundleFromMap(parameters)
            appEventsLogger.logEvent(eventName, parameterBundle)
        } else {
            appEventsLogger.logEvent(eventName)
        }

        result.success(null)
    }

    fun pushNotificationOpen(
        appEventsLogger: AppEventsLogger,
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        val action = call.argument("action") as? String
        val payload = call.argument("payload") as? Map<String, Any>
        val payloadBundle = createBundleFromMap(payload)!!

        if (action != null) {
            appEventsLogger.logPushNotificationOpen(payloadBundle, action)
        } else {
            appEventsLogger.logPushNotificationOpen(payloadBundle)
        }

        result.success(null)
    }

    fun setUserId(call: MethodCall, result: MethodChannel.Result) {
        val id = call.arguments as String
        AppEventsLogger.setUserID(id)
        result.success(null)
    }

    private fun createBundleFromMap(parameterMap: Map<String, Any>?): Bundle? {
        if (parameterMap == null) {
            return null
        }

        val bundle = Bundle()
        for (jsonParam in parameterMap.entries) {
            val value = jsonParam.value
            val key = jsonParam.key
            when (value) {
                is String -> {
                    bundle.putString(key, value)
                }
                is Int -> {
                    bundle.putInt(key, value)
                }
                is Long -> {
                    bundle.putLong(key, value)
                }
                is Double -> {
                    bundle.putDouble(key, value)
                }
                is Boolean -> {
                    bundle.putBoolean(key, value)
                }
                is Map<*, *> -> {
                    val nestedBundle = createBundleFromMap(value as Map<String, Any>)
                    bundle.putBundle(key, nestedBundle as Bundle)
                }
                else -> {
                    throw IllegalArgumentException(
                        "Unsupported value type: " + value.javaClass.kotlin
                    )
                }
            }
        }
        return bundle
    }

    fun setAutoLogAppEventsEnabled(call: MethodCall, result: MethodChannel.Result) {
        val enabled = call.arguments as Boolean
        FacebookSdk.setAutoLogAppEventsEnabled(enabled)
        result.success(null)
    }

    fun setDataProcessingOptions(call: MethodCall, result: MethodChannel.Result) {
        val options = call.argument("options") as? ArrayList<String> ?: arrayListOf()
        val country = call.argument("country") as? Int ?: 0
        val state = call.argument("state") as? Int ?: 0

        FacebookSdk.setDataProcessingOptions(options.toTypedArray(), country, state)
        result.success(null)
    }

    fun purchased(
        appEventsLogger: AppEventsLogger,
        call: MethodCall,
        result: MethodChannel.Result
    ) {
        val amount = (call.argument("amount") as? Double)?.toBigDecimal()
        val currency = Currency.getInstance(call.argument("currency") as? String)
        val parameters = call.argument("parameters") as? Map<String, Any>
        val parameterBundle = createBundleFromMap(parameters) ?: Bundle()

        appEventsLogger.logPurchase(amount, currency, parameterBundle)
        result.success(null)
    }
}

File: ios/Classes/FlutterMetaSdkMethods.swift
//
//  FlutterMetaSdkMethods.swift
//  flutter_meta_sdk
//
//  Created by Chandra Abdul Fattah on 25/07/22.
//

import Foundation
import FBSDKCoreKit
import FBSDKCoreKit_Basics
import FBAudienceNetwork

class FlutterMetaSdkMethods {
    static func clearUserData(result: @escaping FlutterResult) {
        AppEvents.shared.clearUserData()
        result(nil)
    }

    static func setUserData(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = call.arguments as? [String: Any] ?? [String: Any]()

        AppEvents.shared.setUserData(arguments["email"] as? String, forType: FBSDKAppEventUserDataType.email)
        AppEvents.shared.setUserData(arguments["firstName"] as? String, forType: FBSDKAppEventUserDataType.firstName)
        AppEvents.shared.setUserData(arguments["lastName"] as? String, forType: FBSDKAppEventUserDataType.lastName)
        AppEvents.shared.setUserData(arguments["phone"] as? String, forType: FBSDKAppEventUserDataType.phone)
        AppEvents.shared.setUserData(arguments["dateOfBirth"] as? String, forType: FBSDKAppEventUserDataType.dateOfBirth)
        AppEvents.shared.setUserData(arguments["gender"] as? String, forType: FBSDKAppEventUserDataType.gender)
        AppEvents.shared.setUserData(arguments["city"] as? String, forType: FBSDKAppEventUserDataType.city)
        AppEvents.shared.setUserData(arguments["state"] as? String, forType: FBSDKAppEventUserDataType.state)
        AppEvents.shared.setUserData(arguments["zip"] as? String, forType: FBSDKAppEventUserDataType.zip)
        AppEvents.shared.setUserData(arguments["country"] as? String, forType: FBSDKAppEventUserDataType.country)

        result(nil)
    }

    static func clearUserID(result: @escaping FlutterResult) {
        AppEvents.shared.userID = nil
        result(nil)
    }

    static func flush(result: @escaping FlutterResult) {
        AppEvents.shared.flush()
        result(nil)
    }

    /// iOS FBSDKCoreKit doesn't expose a getSdkVersion [like Android does](https://developers.facebook.com/docs/reference/androidsdk/current/facebook/com/facebook/facebooksdk.html/#getsdkversion)
    /// so just return "latest"  as this is what is set in the podspec
    static func getSdkVersion(result: @escaping FlutterResult) {
        result("latest")
    }

    static func getApplicationId(result: @escaping FlutterResult) {
        result(Settings.shared.appID)
    }

    static func getAnonymousId(result: @escaping FlutterResult) {
        result(AppEvents.shared.anonymousID)
    }

    static func logEvent(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = call.arguments as? [String: Any] ?? [String: Any]()
        let eventName = arguments["name"] as! String
        let parameters = arguments["parameters"] as? [AppEvents.ParameterName: Any] ?? [AppEvents.ParameterName: Any]()
        if arguments["_valueToSum"] != nil && !(arguments["_valueToSum"] is NSNull) {
            let valueToDouble = arguments["_valueToSum"] as! Double
            AppEvents.shared.logEvent(AppEvents.Name(eventName), valueToSum: valueToDouble, parameters: parameters)
        } else {
            AppEvents.shared.logEvent(AppEvents.Name(eventName), parameters: parameters)
        }

        result(nil)
    }

    static func pushNotificationOpen(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = call.arguments as? [String: Any] ?? [String: Any]()
        let payload = arguments["payload"] as? [String: Any]
        if let action = arguments["action"] {
            let actionString = action as! String
            AppEvents.shared.logPushNotificationOpen(payload: payload!, action: actionString)
        } else {
            AppEvents.shared.logPushNotificationOpen(payload: payload!)
        }

        result(nil)
    }

    static func setUserId(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let id = call.arguments as! String
        AppEvents.shared.userID = id
        result(nil)
    }

    static func setAutoLogAppEventsEnabled(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let enabled = call.arguments as! Bool
        Settings.shared.isAutoLogAppEventsEnabled = enabled
        result(nil)
    }

    static func setDataProcessingOptions(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = call.arguments as? [String: Any] ?? [String: Any]()
        let modes = arguments["options"] as? [String] ?? []
        let state = arguments["state"] as? Int32 ?? 0
        let country = arguments["country"] as? Int32 ?? 0

        Settings.shared.setDataProcessingOptions(modes, country: country, state: state)

        result(nil)
    }

    static func purchased(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = call.arguments as? [String: Any] ?? [String: Any]()
        let amount = arguments["amount"] as! Double
        let currency = arguments["currency"] as! String
        let parameters = arguments["parameters"] as? [AppEvents.ParameterName: Any] ?? [AppEvents.ParameterName: Any]()
        AppEvents.shared.logPurchase(amount: amount, currency: currency, parameters: parameters)

        result(nil)
    }

    static func setAdvertiserTracking(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        let arguments = call.arguments as? [String: Any] ?? [String: Any]()
        let enabled = arguments["enabled"] as! Bool
        let collectId = arguments["collectId"] as! Bool
        Settings.shared.isAdvertiserTrackingEnabled = enabled
        Settings.shared.isAdvertiserIDCollectionEnabled = collectId
        result(nil)
    }
    
    static func activateApp(){
        AppEvents.shared.activateApp()
    }
}
